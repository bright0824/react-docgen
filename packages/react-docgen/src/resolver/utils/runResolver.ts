import type FileState from '../../FileState.js';
import type { Resolver, ResolverClass, ResolverFunction } from '../index.js';

function isResolverClass(resolver: Resolver): resolver is ResolverClass {
  return typeof resolver === 'object';
}

export default function runResolver(
  resolver: Resolver,
  file: FileState,
): ReturnType<ResolverFunction> {
  return isResolverClass(resolver) ? resolver.resolve(file) : resolver(file);
}
