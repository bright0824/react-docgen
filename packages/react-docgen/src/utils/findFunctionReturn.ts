import type { NodePath } from '@babel/traverse';
import { visitors } from '@babel/traverse';
import resolveToValue from './resolveToValue';
import { ignore } from './traverse';

type Predicate = (p: NodePath) => boolean;
interface TraverseState {
  readonly predicate: Predicate;
  resolvedReturnPath?: NodePath;
  readonly seen: WeakSet<NodePath>;
}

const explodedVisitors = visitors.explode<TraverseState>({
  Function: { enter: ignore },
  Class: { enter: ignore },
  ObjectExpression: { enter: ignore },
  ReturnStatement: {
    enter: function (path, state) {
      const argument = path.get('argument');

      if (argument.hasNode()) {
        const resolvedPath = resolvesToFinalValue(
          argument,
          state.predicate,
          state.seen,
        );

        if (resolvedPath) {
          state.resolvedReturnPath = resolvedPath;
          path.stop();
        }
      }
    },
  },
});

function resolvesToFinalValue(
  path: NodePath,
  predicate: Predicate,
  seen: WeakSet<NodePath>,
): NodePath | undefined {
  // avoid returns with recursive function calls
  if (seen.has(path)) {
    return;
  }
  seen.add(path);

  // Is the path already passes then return it.
  if (predicate(path)) {
    return path;
  }

  const resolvedPath = resolveToValue(path);

  // If the resolved path is already passing then no need to further check
  // Only do this if the resolvedPath actually resolved something as otherwise we did this check already
  if (resolvedPath.node !== path.node && predicate(resolvedPath)) {
    return resolvedPath;
  }

  // If the path points to a conditional expression, then we need to look only at
  // the two possible paths
  if (resolvedPath.isConditionalExpression()) {
    return (
      resolvesToFinalValue(resolvedPath.get('consequent'), predicate, seen) ||
      resolvesToFinalValue(resolvedPath.get('alternate'), predicate, seen)
    );
  }

  // If the path points to a logical expression (AND, OR, ...), then we need to look only at
  // the two possible paths
  if (resolvedPath.isLogicalExpression()) {
    return (
      resolvesToFinalValue(resolvedPath.get('left'), predicate, seen) ||
      resolvesToFinalValue(resolvedPath.get('right'), predicate, seen)
    );
  }

  // If we have a call expression, lets try to follow it
  if (resolvedPath.isCallExpression()) {
    const returnValue = findFunctionReturnWithCache(
      resolveToValue(resolvedPath.get('callee')),
      predicate,
      seen,
    );

    if (returnValue) {
      return returnValue;
    }
  }

  return;
}

/**
 * This can be used in two ways
 * 1. Find the first return path that passes the predicate function
 *    (for example to check if a function is returning something)
 * 2. Find all occurrences of return values
 *    For this the predicate acts more like a collector and always needs to return false
 */
function findFunctionReturnWithCache(
  path: NodePath,
  predicate: Predicate,
  seen: WeakSet<NodePath>,
): NodePath | undefined {
  let functionPath: NodePath = path;

  if (functionPath.isObjectProperty()) {
    functionPath = functionPath.get('value');
  } else if (functionPath.isClassProperty()) {
    const classPropertyValue = functionPath.get('value');

    if (classPropertyValue.hasNode()) {
      functionPath = classPropertyValue;
    }
  }

  if (!functionPath.isFunction()) {
    return;
  }

  // skip traversing for ArrowFunctionExpressions with no block
  if (path.isArrowFunctionExpression()) {
    const body = path.get('body');

    if (!body.isBlockStatement()) {
      return resolvesToFinalValue(body, predicate, seen);
    }
  }

  const state: TraverseState = {
    predicate,
    seen,
  };

  path.traverse(explodedVisitors, state);

  return state.resolvedReturnPath;
}

/**
 * This can be used in two ways
 * 1. Find the first return path that passes the predicate function
 *    (for example to check if a function is returning something)
 * 2. Find all occurrences of return values
 *    For this the predicate acts more like a collector and always needs to return false
 */
export default function findFunctionReturn(
  path: NodePath,
  predicate: Predicate,
): NodePath | undefined {
  return findFunctionReturnWithCache(path, predicate, new WeakSet());
}
