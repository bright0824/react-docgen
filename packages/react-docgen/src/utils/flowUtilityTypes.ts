import type { NodePath } from '@babel/traverse';
import type { GenericTypeAnnotation } from '@babel/types';

/**
 * See `supportedUtilityTypes` for which types are supported and
 * https://flow.org/en/docs/types/utilities/ for which types are available.
 */
export function isSupportedUtilityType(
  path: NodePath,
): path is NodePath<GenericTypeAnnotation> {
  if (path.isGenericTypeAnnotation()) {
    const idPath = path.get('id');

    if (idPath.isIdentifier()) {
      const name = idPath.node.name;

      return name === '$Exact' || name === '$ReadOnly';
    }
  }

  return false;
}

/**
 * Unwraps well known utility types. For example:
 *
 *   $ReadOnly<T> => T
 */
export function unwrapUtilityType(path: NodePath): NodePath {
  let resultPath: NodePath = path;
  while (isSupportedUtilityType(resultPath)) {
    const typeParameters = resultPath.get('typeParameters');
    if (!typeParameters.hasNode()) break;

    const firstParam = typeParameters.get('params')[0];
    if (!firstParam) break;

    resultPath = firstParam;
  }

  return resultPath;
}
