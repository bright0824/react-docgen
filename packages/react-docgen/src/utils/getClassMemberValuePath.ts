import type { NodePath } from '@babel/traverse';
import type {
  ClassDeclaration,
  ClassExpression,
  ClassMethod,
  ClassProperty,
  Expression,
} from '@babel/types';
import getNameOrValue from './getNameOrValue';

export default function getClassMemberValuePath(
  classDefinition: NodePath<ClassDeclaration | ClassExpression>,
  memberName: string,
): NodePath<ClassMethod | Expression> | null {
  const classMember = classDefinition
    .get('body')
    .get('body')
    .find(memberPath => {
      if (
        (memberPath.isClassMethod() && memberPath.node.kind !== 'set') ||
        memberPath.isClassProperty()
      ) {
        const key = (memberPath as NodePath<ClassMethod | ClassProperty>).get(
          'key',
        );

        return (
          (!memberPath.node.computed || key.isLiteral()) &&
          getNameOrValue(key) === memberName
        );
      }

      return false;
    });

  if (classMember) {
    // For ClassProperty we return the value and for ClassMethod
    // we return itself
    return classMember.isClassMethod()
      ? classMember
      : (classMember.get('value') as NodePath<Expression>);
  }

  return null;
}
