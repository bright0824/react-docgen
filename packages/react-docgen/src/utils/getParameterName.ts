import type { NodePath } from '@babel/traverse';
import type {
  ArrayPattern,
  AssignmentPattern,
  Identifier,
  ObjectPattern,
  RestElement,
  TSParameterProperty,
} from '@babel/types';
import printValue from './printValue.js';

type ParameterNodePath = NodePath<
  | ArrayPattern
  | AssignmentPattern
  | Identifier
  | ObjectPattern
  | RestElement
  | TSParameterProperty
>;

export default function getParameterName(
  parameterPath: ParameterNodePath,
): string {
  if (parameterPath.isIdentifier()) {
    return parameterPath.node.name;
  } else if (parameterPath.isAssignmentPattern()) {
    return getParameterName(parameterPath.get('left') as ParameterNodePath);
  } else if (
    parameterPath.isObjectPattern() ||
    parameterPath.isArrayPattern()
  ) {
    return printValue(parameterPath);
  } else if (parameterPath.isRestElement()) {
    return `...${getParameterName(
      parameterPath.get('argument') as ParameterNodePath,
    )}`;
  } else if (parameterPath.isTSParameterProperty()) {
    return getParameterName(parameterPath.get('parameter'));
  }

  throw new TypeError(
    'Parameter name must be one of Identifier, AssignmentPattern, ArrayPattern, ' +
      `ObjectPattern or RestElement, instead got ${parameterPath.node.type}`,
  );
}
