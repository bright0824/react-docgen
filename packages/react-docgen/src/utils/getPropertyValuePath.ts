import type { NodePath } from '@babel/traverse';
import type { Expression, ObjectExpression, ObjectMethod } from '@babel/types';
import getPropertyName from './getPropertyName.js';

/**
 * Given an ObjectExpression, this function returns the path of the value of
 * the property with name `propertyName`. if the property is an ObjectMethod we
 * return the ObjectMethod itself.
 */
export default function getPropertyValuePath(
  path: NodePath<ObjectExpression>,
  propertyName: string,
): NodePath<Expression | ObjectMethod> | null {
  const property = path
    .get('properties')
    .find(
      propertyPath =>
        !propertyPath.isSpreadElement() &&
        getPropertyName(propertyPath) === propertyName,
    );

  if (property) {
    return property.isObjectMethod()
      ? property
      : (property.get('value') as NodePath<Expression>);
  }

  return null;
}
