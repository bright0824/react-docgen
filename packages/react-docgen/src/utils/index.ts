import * as docblock from './docblock';
import * as expressionTo from './expressionTo';
import * as flowUtilityTypes from './flowUtilityTypes';
import * as traverse from './traverse';

export { docblock, expressionTo, flowUtilityTypes, traverse };

export { default as getClassMemberValuePath } from './getClassMemberValuePath';
export { default as getFlowType } from './getFlowType';
export { default as getMemberExpressionRoot } from './getMemberExpressionRoot';
export { default as getMemberExpressionValuePath } from './getMemberExpressionValuePath';
export { default as getMembers } from './getMembers';
export {
  default as getMemberValuePath,
  isSupportedDefinitionType,
} from './getMemberValuePath';
export { default as getMethodDocumentation } from './getMethodDocumentation';
export type { MethodNodePath } from './getMethodDocumentation';
export { default as getNameOrValue } from './getNameOrValue';
export { default as getParameterName } from './getParameterName';
export { default as getPropertyName, COMPUTED_PREFIX } from './getPropertyName';
export { default as getPropertyValuePath } from './getPropertyValuePath';
export { default as getPropType } from './getPropType';
export { default as getTSType } from './getTSType';
export { default as getTypeAnnotation } from './getTypeAnnotation';
export {
  default as getTypeFromReactComponent,
  applyToTypeProperties,
} from './getTypeFromReactComponent';
export { default as getTypeIdentifier } from './getTypeIdentifier';
export { default as getTypeParameters } from './getTypeParameters';
export type { TypeParameters } from './getTypeParameters';
export { default as isDestructuringAssignment } from './isDestructuringAssignment';
export { default as isExportsOrModuleAssignment } from './isExportsOrModuleAssignment';
export { default as isReactBuiltinCall } from './isReactBuiltinCall';
export { default as isReactChildrenElementCall } from './isReactChildrenElementCall';
export { default as isReactCloneElementCall } from './isReactCloneElementCall';
export { default as isReactComponentClass } from './isReactComponentClass';
export { default as isReactComponentMethod } from './isReactComponentMethod';
export { default as isReactCreateClassCall } from './isReactCreateClassCall';
export { default as isReactCreateElementCall } from './isReactCreateElementCall';
export { default as isReactForwardRefCall } from './isReactForwardRefCall';
export { default as isReactModuleName } from './isReactModuleName';
export { default as isRequiredPropType } from './isRequiredPropType';
export { default as isStatelessComponent } from './isStatelessComponent';
export { default as isUnreachableFlowType } from './isUnreachableFlowType';
export { default as match } from './match';
export { default as normalizeClassDefinition } from './normalizeClassDefinition';
export { default as parseJsDoc } from './parseJsDoc';
export { default as postProcessDocumentation } from './postProcessDocumentation';
export { default as printValue } from './printValue';
export { default as resolveExportDeclaration } from './resolveExportDeclaration';
export { default as resolveFunctionDefinitionToReturnValue } from './resolveFunctionDefinitionToReturnValue';
export { default as resolveGenericTypeAnnotation } from './resolveGenericTypeAnnotation';
export { default as resolveHOC } from './resolveHOC';
export { default as resolveObjectPatternPropertyToValue } from './resolveObjectPatternPropertyToValue';
export {
  default as resolveObjectKeysToArray,
  resolveObjectToNameArray,
} from './resolveObjectKeysToArray';
export { default as resolveObjectValuesToArray } from './resolveObjectValuesToArray';
export { default as resolveToModule } from './resolveToModule';
export { default as resolveToValue } from './resolveToValue';
export { default as setPropDescription } from './setPropDescription';
