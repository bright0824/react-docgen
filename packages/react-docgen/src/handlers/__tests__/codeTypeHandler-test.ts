import { parse, makeMockImporter } from '../../../tests/utils';
import Documentation from '../../Documentation';
import codeTypeHandler from '../codeTypeHandler.js';
import type DocumentationMock from '../../__mocks__/Documentation';
import type {
  ArrowFunctionExpression,
  CallExpression,
  ClassDeclaration,
  ExportNamedDeclaration,
  ObjectExpression,
} from '@babel/types';
import type { NodePath } from '@babel/traverse';
import type { ComponentNode } from '../../resolver';
import { beforeEach, describe, expect, test, vi } from 'vitest';

vi.mock('../../Documentation.js');
vi.mock('../../utils/getFlowType.js', () => ({
  default: () => ({}),
  __esModule: true,
}));

describe('codeTypeHandler', () => {
  let documentation: Documentation & DocumentationMock;

  beforeEach(() => {
    documentation = new Documentation() as Documentation & DocumentationMock;
  });

  const mockImporter = makeMockImporter({
    something: stmtLast =>
      stmtLast<ExportNamedDeclaration>(`
      export type Props = {
        foo: string,
        bar?: number,
        hal: boolean,
        [key: string]: string,
        abc: string | number,
        def: "test" | 1 | true,
        foobar: Foo & Bar,
      };
    `).get('declaration') as NodePath,
  });

  function template(src: string, typeObject: string): string {
    return `
      ${src}
      var React = require('React');
      var Component = React.Component;

      type Props = ${typeObject};
    `;
  }

  function testCodeTypeHandler(
    getSrc: (src: string) => NodePath<ComponentNode>,
  ) {
    test('detects types correctly', () => {
      const flowTypesSrc = `
      {
        foo: string,
        bar: number,
        hal: boolean,
      }
      `;
      const definition = getSrc(flowTypesSrc);

      codeTypeHandler(documentation, definition);

      expect(documentation.descriptors).toEqual({
        foo: {
          flowType: {},
          required: true,
          description: '',
        },
        bar: {
          flowType: {},
          required: true,
          description: '',
        },
        hal: {
          flowType: {},
          required: true,
          description: '',
        },
      });
    });

    test('detects whether a prop is required', () => {
      const flowTypesSrc = `
      {
        foo: string,
        bar?: number,
      }
      `;
      const definition = getSrc(flowTypesSrc);

      codeTypeHandler(documentation, definition);

      expect(documentation.descriptors).toEqual({
        foo: {
          flowType: {},
          required: true,
          description: '',
        },
        bar: {
          flowType: {},
          required: false,
          description: '',
        },
      });
    });

    test('ignores hash map entry', () => {
      const flowTypesSrc = `
      {
        [key: string]: string,
        bar?: number,
      }
      `;
      const definition = getSrc(flowTypesSrc);

      codeTypeHandler(documentation, definition);

      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('detects union types', () => {
      const flowTypesSrc = `
      {
        foo: string | number,
        bar: "test" | 1 | true,
      }
      `;
      const definition = getSrc(flowTypesSrc);

      codeTypeHandler(documentation, definition);

      expect(documentation.descriptors).toEqual({
        foo: {
          flowType: {},
          required: true,
          description: '',
        },
        bar: {
          flowType: {},
          required: true,
          description: '',
        },
      });
    });

    test('detects intersection types', () => {
      const flowTypesSrc = `
      {
        foo: Foo & Bar,
      }
      `;
      const definition = getSrc(flowTypesSrc);

      codeTypeHandler(documentation, definition);

      expect(documentation.descriptors).toEqual({
        foo: {
          flowType: {},
          required: true,
          description: '',
        },
      });
    });

    describe('special generic type annotations', () => {
      ['$ReadOnly', '$Exact'].forEach(annotation => {
        test(`unwraps ${annotation}<...>`, () => {
          const flowTypesSrc = `
            ${annotation}<{
              foo: string | number,
            }>
          `;

          const definition = getSrc(flowTypesSrc);

          codeTypeHandler(documentation, definition);

          expect(documentation.descriptors).toEqual({
            foo: {
              flowType: {},
              required: true,
              description: '',
            },
          });
        });
      });
    });
  }

  describe('TypeAlias', () => {
    describe('class definition for flow <0.53', () => {
      testCodeTypeHandler(propTypesSrc =>
        parse.statement(
          template(
            'class Foo extends Component<void, Props, void> {}',
            propTypesSrc,
          ),
        ),
      );
    });

    describe('class definition for flow >=0.53 without State', () => {
      testCodeTypeHandler(propTypesSrc =>
        parse.statement(
          template('class Foo extends Component<Props> {}', propTypesSrc),
        ),
      );
    });

    describe('class definition for flow >=0.53 with State', () => {
      testCodeTypeHandler(propTypesSrc =>
        parse.statement(
          template(
            'class Foo extends Component<Props, State> {}',
            propTypesSrc,
          ),
        ),
      );
    });

    describe('class definition with inline props', () => {
      testCodeTypeHandler(propTypesSrc =>
        parse.statement(
          template(
            'class Foo extends Component { props: Props; }',
            propTypesSrc,
          ),
        ),
      );
    });

    describe('stateless component', () => {
      testCodeTypeHandler(
        propTypesSrc =>
          parse
            .statement(template('(props: Props) => <div />;', propTypesSrc))
            .get('expression') as NodePath<ArrowFunctionExpression>,
      );
    });
  });

  describe('does not error if flowTypes cannot be found', () => {
    test('ObjectExpression', () => {
      const definition = parse.expression<ObjectExpression>('{fooBar: 42}');

      expect(() => codeTypeHandler(documentation, definition)).not.toThrow();
    });

    test('ClassDeclaration', () => {
      const definition = parse.statement<ClassDeclaration>(
        'class Foo extends Component {}',
      );

      expect(() => codeTypeHandler(documentation, definition)).not.toThrow();
    });

    test('ArrowFunctionExpression', () => {
      const definition =
        parse.statement<ArrowFunctionExpression>('() => <div />');

      expect(() => codeTypeHandler(documentation, definition)).not.toThrow();
    });
  });

  test('supports intersection proptypes', () => {
    const definition = parse
      .statement(
        `(props: Props) => <div />;
         var React = require('React');
         import type Imported from 'something';
         type Props = Imported & { foo: 'bar' };`,
      )
      .get('expression') as NodePath<ArrowFunctionExpression>;

    codeTypeHandler(documentation, definition);

    expect(documentation.descriptors).toEqual({
      foo: {
        flowType: {},
        required: true,
        description: '',
      },
    });
  });

  test('does support utility types inline', () => {
    const definition = parse
      .statement(
        `(props: $ReadOnly<Props>) => <div />;
         var React = require('React');
         type Props = { foo: 'fooValue' };`,
      )
      .get('expression') as NodePath<ArrowFunctionExpression>;

    expect(() => codeTypeHandler(documentation, definition)).not.toThrow();
    expect(documentation.descriptors).toMatchSnapshot();
  });

  test('does not support union proptypes', () => {
    const definition = parse
      .statement(
        `(props: Props) => <div />;
         var React = require('React');
         import type Imported from 'something';
         type Other = { bar: 'barValue' };
         type Props = Imported | Other | { foo: 'fooValue' };`,
      )
      .get('expression') as NodePath<ArrowFunctionExpression>;

    expect(() => codeTypeHandler(documentation, definition)).not.toThrow();
    expect(documentation.descriptors).toEqual({});
  });

  describe('imported prop types', () => {
    test('does not resolve type included by require', () => {
      const definition = parse
        .statement(
          `(props: Props) => <div />;
           var React = require('React');
           var Component = React.Component;
           var Props = require('something');`,
          mockImporter,
        )
        .get('expression') as NodePath<ArrowFunctionExpression>;

      expect(() => codeTypeHandler(documentation, definition)).not.toThrow();
      expect(documentation.descriptors).toEqual({});
    });

    test('imported', () => {
      const definition = parse
        .statement(
          `(props: Props) => <div />;
           var React = require('React');
           var Component = React.Component;
           import { Props } from 'something';`,
          mockImporter,
        )
        .get('expression') as NodePath<ArrowFunctionExpression>;

      expect(() => codeTypeHandler(documentation, definition)).not.toThrow();
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('type not imported', () => {
      const definition = parse
        .statement(
          `(props: Props) => <div />;
           var React = require('React');
           var Component = React.Component;
           import type { Props } from 'something';`,
        )
        .get('expression') as NodePath<ArrowFunctionExpression>;

      expect(() => codeTypeHandler(documentation, definition)).not.toThrow();
      expect(documentation.descriptors).toEqual({});
    });

    test('type imported', () => {
      const definition = parse
        .statement(
          `(props: Props) => <div />;
           var React = require('React');
           var Component = React.Component;
           import type { Props } from 'something';`,
          mockImporter,
        )
        .get('expression') as NodePath<ArrowFunctionExpression>;

      expect(() => codeTypeHandler(documentation, definition)).not.toThrow();
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('does not resolve types not in scope', () => {
      const definition = parse
        .statement(
          `(props: Props) => <div />;
           var React = require('React');
           var Component = React.Component;`,
          mockImporter,
        )
        .get('expression') as NodePath<ArrowFunctionExpression>;

      expect(() => codeTypeHandler(documentation, definition)).not.toThrow();
      expect(documentation.descriptors).toEqual({});
    });

    test('does not resolve types not in scope', () => {
      const definition = parse
        .statement(
          `(props: Props) => <div />;
           var React = require('React');
           var Component = React.Component;`,
          mockImporter,
        )
        .get('expression') as NodePath<ArrowFunctionExpression>;

      expect(() => codeTypeHandler(documentation, definition)).not.toThrow();
      expect(documentation.descriptors).toEqual({});
    });
  });

  describe('forwardRef', () => {
    test('resolves prop type from function expression', () => {
      const src = `
        import React from 'react';
        type Props = { foo: string };
        React.forwardRef((props: Props, ref) => <div ref={ref}>{props.foo}</div>);
      `;

      codeTypeHandler(documentation, parse.expressionLast<CallExpression>(src));
      expect(documentation.descriptors).toEqual({
        foo: {
          flowType: {},
          required: true,
          description: '',
        },
      });
    });

    test('resolves when the function is not inline', () => {
      const src = `
        import React from 'react';
        type Props = { foo: string };
        const ComponentImpl = (props: Props, ref) => <div ref={ref}>{props.foo}</div>;
        React.forwardRef(ComponentImpl);
      `;

      codeTypeHandler(documentation, parse.expressionLast<CallExpression>(src));
      expect(documentation.descriptors).toEqual({
        foo: {
          flowType: {},
          required: true,
          description: '',
        },
      });
    });

    test('resolves when the function is rebound and not inline', () => {
      const src = `
        import React from 'react';
        type Props = { foo: string };
        let Component = (props: Props, ref) => <div ref={ref}>{props.foo}</div>;
        Component = React.forwardRef(Component);
      `;

      codeTypeHandler(
        documentation,
        parse.expressionLast(src).get('right') as NodePath<CallExpression>,
      );
      expect(documentation.descriptors).toEqual({
        foo: {
          flowType: {},
          required: true,
          description: '',
        },
      });
    });
  });
});
