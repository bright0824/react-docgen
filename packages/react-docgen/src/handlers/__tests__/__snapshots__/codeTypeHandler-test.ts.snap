// Vitest Snapshot v1

exports[`codeTypeHandler > TypeAlias > class definition for flow <0.53 > ignores hash map entry 1`] = `
{
  "bar": {
    "description": "",
    "flowType": {},
    "required": false,
  },
}
`;

exports[`codeTypeHandler > TypeAlias > class definition for flow >=0.53 with State > ignores hash map entry 1`] = `
{
  "bar": {
    "description": "",
    "flowType": {},
    "required": false,
  },
}
`;

exports[`codeTypeHandler > TypeAlias > class definition for flow >=0.53 without State > ignores hash map entry 1`] = `
{
  "bar": {
    "description": "",
    "flowType": {},
    "required": false,
  },
}
`;

exports[`codeTypeHandler > TypeAlias > class definition with inline props > ignores hash map entry 1`] = `
{
  "bar": {
    "description": "",
    "flowType": {},
    "required": false,
  },
}
`;

exports[`codeTypeHandler > TypeAlias > stateless component > ignores hash map entry 1`] = `
{
  "bar": {
    "description": "",
    "flowType": {},
    "required": false,
  },
}
`;

exports[`codeTypeHandler > does support utility types inline 1`] = `
{
  "foo": {
    "description": "",
    "flowType": {},
    "required": true,
  },
}
`;

exports[`codeTypeHandler > imported prop types > imported 1`] = `
{
  "abc": {
    "description": "",
    "flowType": {},
    "required": true,
  },
  "bar": {
    "description": "",
    "flowType": {},
    "required": false,
  },
  "def": {
    "description": "",
    "flowType": {},
    "required": true,
  },
  "foo": {
    "description": "",
    "flowType": {},
    "required": true,
  },
  "foobar": {
    "description": "",
    "flowType": {},
    "required": true,
  },
  "hal": {
    "description": "",
    "flowType": {},
    "required": true,
  },
}
`;

exports[`codeTypeHandler > imported prop types > type imported 1`] = `
{
  "abc": {
    "description": "",
    "flowType": {},
    "required": true,
  },
  "bar": {
    "description": "",
    "flowType": {},
    "required": false,
  },
  "def": {
    "description": "",
    "flowType": {},
    "required": true,
  },
  "foo": {
    "description": "",
    "flowType": {},
    "required": true,
  },
  "foobar": {
    "description": "",
    "flowType": {},
    "required": true,
  },
  "hal": {
    "description": "",
    "flowType": {},
    "required": true,
  },
}
`;
